In [7]:
def factorial(n):
    if n == 1 or n == 0:  # Base case
        return 1
    return n * factorial(n - 1)  # Recursive case

factorial(5)

120

Let's perform a **dry run** of the `factorial(n)` function for `n = 5`.  

---

### **Given function:**
```python
def factorial(n):
    if n == 1 or n == 0:  # Base case
        return 1
    return n * factorial(n - 1)  # Recursive case
```

---

### **Step-by-step dry run for `factorial(5)`:**

1. **First Call:** `factorial(5)`  
   - Condition: `n == 1 or n == 0` → False  
   - Returns: `5 * factorial(4)`

2. **Second Call:** `factorial(4)`  
   - Condition: `n == 1 or n == 0` → False  
   - Returns: `4 * factorial(3)`

3. **Third Call:** `factorial(3)`  
   - Condition: `n == 1 or n == 0` → False  
   - Returns: `3 * factorial(2)`

4. **Fourth Call:** `factorial(2)`  
   - Condition: `n == 1 or n == 0` → False  
   - Returns: `2 * factorial(1)`

5. **Fifth Call:** `factorial(1)`  
   - Condition: `n == 1 or n == 0` → True (Base case)  
   - Returns: `1` (Stops recursion)

---

### **Returning values step-by-step:**  

Now, the recursion starts **unwinding**, returning values from the deepest call back to the top:  

- `factorial(1)` → `1`  
- `factorial(2)` → `2 * 1 = 2`  
- `factorial(3)` → `3 * 2 = 6`  
- `factorial(4)` → `4 * 6 = 24`  
- `factorial(5)` → `5 * 24 = 120`  

---

### **Final Answer:**  
`factorial(5) = 120`

---

Let me know if you have any questions! 😊